modules/dnn/src/net_impl2.cpp

```diff
@@ -0,0 +1,1105 @@
+                    value, dimnames_[idx].c_str(), value0));
+        }
+    }
+}*/
+
+void Net::Impl::traceArg(std::ostream& strm_, const char* prefix, size_t i, Arg arg, bool dumpdata)
+{
+    const int PPRINT_CONTEXT = 3;
+    const int PPRINT_CONST_THRESHOLD = 16;
+    const int PPRINT_ALL_THRESHOLD = 100;
+    const Mat& m = argTensor(arg);
+    const ArgData& adata = args.at(arg.idx);
+    bool constArg = adata.kind == DNN_ARG_CONST;
+    // [TODO] replace with type compatibility check
+    // CV_Assert(m.type() == adata.type);
+    strm_ << prefix << " " << i << ". Name: " << (arg.idx > 0 ? adata.name.c_str() : "<empty>") << "\n";
+    if (arg.idx == 0)
+        return;
+    strm_ << "  Buf: " << bufidxs.at(arg.idx) << "\n";
+    strm_ << "  Type: " << typeToString(adata.type) << " \n";
+    MatShape shape = m.shape();
+    strm_ << "  Shape: " << shape;
+    if (constArg && m.total() <= PPRINT_CONST_THRESHOLD) {
+        strm_ << " /* ";
+        pprint(strm_, m, 0, PPRINT_CONTEXT, PPRINT_CONST_THRESHOLD, '{');
+        strm_ << " */";
+    }
+    strm_ << "\n  Layout: " << layoutToString(shape.layout) << "\n";
+    if (dumpdata && !constArg) {
+        // [TODO] when we support block layout, block-layout tensor
+        // should be converted to the original layout before printing it
+        pprint(strm_, m, 0, PPRINT_CONTEXT, PPRINT_ALL_THRESHOLD, '[');
+        strm_ << "\n";
+    }
+}
+
+void Net::Impl::setMainGraphInput(InputArray m, const std::string& inpname)
+{
+    CV_Assert(mainGraph);
+    const std::vector<Arg>& gr_inputs = mainGraph->inputs();
+    size_t i, ninputs = gr_inputs.size();
+    if (inpname.empty()) {
+        CV_Assert(ninputs == 1 && "empty name can only be used to set input if there is just one input");
+        i = 0;
+    } else {
+        for (i = 0; i < ninputs; i++) {
+            const ArgData& adata = args.at(gr_inputs[i].idx);
+            CV_Assert(adata.kind == DNN_ARG_INPUT);
+            if (adata.name == inpname)
+                break;
+        }
+        if ((i == ninputs) && (!isdigit(inpname[0]) || !sscanf(inpname.c_str(), "%zu", &i))) {
+            CV_Error_(Error::StsObjectNotFound, ("input '%s' is not found", inpname.c_str()));
+        }
+    }
+    setGraphInput(mainGraph, i, m.getMat());
+}
+
+void Net::Impl::setGraphInput(Ptr<Graph>& graph, size_t idx, const Mat& m)
+{
+    int mtype = m.type();
+    MatShape mshape = m.shape();
+    const std::vector<Arg>& gr_inputs = graph->inputs();
+    CV_Assert(idx < gr_inputs.size());
+    Arg inp = gr_inputs[idx];
+    const ArgData& adata = args.at(inp.idx);
+    /*
+     [TODO] add more detailed shape check
+     if (adata.shape.dims != mshape.dims) {
+     CV_Error_(Error::StsBadArg, ("wrong dimensionality of argument '%s': %d given, %d expected",
+     adata.name.c_str(), tsize.ndims, adata.size.ndims));
+     }
+
+     for (int k = 0; k < mshape.dims; k++) {
+        checkAndUpdateDim(graph, Node(), inp, k, tsize.size[k]);
+     }
+    */
```

Comment:
Need to be fixed before merge.

Reply:
disagree. It's quite difficult to fix properly. It will surely breaks some test cases that just ignore shapes in the model and pass tensors with completely different shapes

Reply:
It means that the tests should be fixed.

Reply:
that's a lot of work. Maybe another 2 weeks. If someone has that much time, please, go ahead. The new engine does no less sanity checks than the old engine and is more robust in general. The commented off code is some extra feature on top of the old engine; it can be polished and enabled in the later PRs after alpha.